### PR TITLE
Pixel projectiles

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
This PR changes projectiles from flying on a terrible precalculated path that we all know and love to using proper collision. It would be easier to show than explain, so see below: 
awxz8ya

Furthermore, this PR changes no projectile sprites and adds sounds for collisions with walls (misses). These projectiles add dynamic spread for automatic (burst) weapons (such as the mech LMG) and shotguns.

Some gifs:
epic
meme

100% original code no ban evasion involved by paprika and no work by @crystalwarrior from hippie either

:cl: Paprika and Crystalwarrior
rscadd: Pixel projectiles. This is easier to see than explain, so go fire a weapon.
/:cl:

### CODE REVIEW

File: code/game/objects/items/stacks/sheets/sheet_types.dm [hunk 1]
```diff
@@ -33,7 +33,7 @@ var/global/list/datum/stack_recipe/metal_recipes = list ( \
 	new/datum/stack_recipe("machine frame", /obj/machinery/constructable_frame/machine_frame, 5, time = 25, one_per_turf = 1, on_floor = 1), \
 	new/datum/stack_recipe("turret frame", /obj/machinery/porta_turret_construct, 5, time = 25, one_per_turf = 1, on_floor = 1), \
 	new/datum/stack_recipe("meatspike frame", /obj/structure/kitchenspike_frame, 5, time = 25, one_per_turf = 1, on_floor = 1), \
-	new/datum/stack_recipe("reflector frame", /obj/structure/reflector, 5, time = 25, one_per_turf = 1, on_floor = 1), \
+/*	new/datum/stack_recipe("reflector frame", /obj/structure/reflector, 5, time = 25, one_per_turf = 1, on_floor = 1), \*/
```
Comment (ChangelingRain) [inline code/game/objects/items/stacks/sheets/sheet_types.dm]: >reflectors /still/ break

Comment (KorPhaeron) [inline code/game/objects/items/stacks/sheets/sheet_types.dm]: Acceptable losses

**Comment by ChangelingRain** [comment]
Am I hearing "unpredictable combat mechanics that don't let you shoot around corners"? I feel like I am!

**Comment by KorPhaeron** [comment]
How is the projectile going exactly where you click along a straight path unpredictable though

**Comment by ChangelingRain** [comment]
It suddenly changes the combat from "can shoot around corners in some directions" to "can't shoot around corners if the projectile path includes the corner you're shooting around"

IE: you can't shoot out of a maint doorway to either side of it, just in front/back of it.
Right now you can shoot to either side and in front/back but projectiles get caught on shots at tiles that are a little to either side of in front/back.

This has been a problem with every attempt to do pixel projectiles.

**Comment by Iamgoofball** [comment]
who cares its prettier

**Comment by PKPenguin321** [comment]
i'm one the ropes with this, one one hand i've seen it in game and it looks admittedly really cool, but on the other hand the unpredictability that changelingrain mentions is still present

**Comment by KazeEspada** [comment]
Have you tried leaving the door way

**Comment by ToadStar** [comment]
Laser Gatling is bugged with this, it won't shoot where you move your cursor to, locking the projectiles into a single straight beam. It looks cool, but I can imagine why that would be a problem.

**Comment by KorPhaeron** [comment]
Okay so I just tested, you can't shoot around corners north/south anyway with our current system, only east/west

**Comment by Fox-McCloud** [comment]
Big :+1: 

Goon and Paradise both have this--IMO, it doesn't make the hitboxes any jankier or anything--just makes things more intuitive.

Players got a bad taste for this the first time it was tried because ti was buggy---provided the bugs/kinks are worked out, this is a very nice system.

**Comment by Impisi** [comment]
I really hope this merges.

Mostly because improvised shells will look and work much better with this system.

**Comment by kevinz000** [comment]
Please, please no.
I rely on bullet curving to robust the lessers.

**Comment by lzimann** [comment]
this looks interesting

**Comment by Supermichael777** [comment]
the terible precalcs need to die

**Comment by Bawhoppen** [comment]
I am conflicted 
I need to see it ingame

**Comment by lzimann** [comment]
it is testmerged in sybil

**Comment by Bawhoppen** [comment]
yes I just got back to my computer

**Comment by KorPhaeron** [comment]
I actually find this much more difficult to aim with for some reason

**Comment by Bawhoppen** [comment]
also
>this
>right as I wanted to add bulletholes

**Comment by Bawhoppen** [comment]
okay after testing :+1:x:100:

**Comment by Robustin** [comment]
3 issues:

1) http://imgur.com/8vofJdE

That's a laser about to hit shaps despite the fact it looks like its going to hit 2 tiles away

2) YOU TOOK DEVIATION OFF MECH WEAPONS THAT WAS MY FAVORITE VAR AND YOU KILLED IT YOU MONSTER PUT IT BACK. 

3) 
>Bawhoppen passing judgment on a PR after wildly blasting at people for 60 seconds on Thunderdome.

**Comment by Bawhoppen** [comment]
It was way more fun than wildly blasting with the old system

**Comment by Durkel** [comment]
I'm finding it very easy to dodge shots that look like they should hit me, and I'm being hit by shit that doesn't look like it should hit me.

There's also the problem that if someones at a angle at a corner, and you fire at them at way that it Looks like it should hit,but it won't hit. You couldn't do that before but it's very misleading now.

**Comment by Incoming5643** [comment]
Incentivizes the less than obvious tactic of aiming your clicks far behind your intended target to get better "control" of the spreading.

Technically punishes players playing ss13 on smaller resolutions.

**Comment by Core0verload** [comment]
It's a really misleading system. Weird collisions is the reason any attempt at making this didn't worked out, and this PR still has those.

**Comment by Incoming5643** [comment]
Honestly it would be less deceptive if the shots WEREN'T angled to show them moving in perfectly straight lines, because mechanically they aren't.

They just partition out horizontal to vertical tiles moves more evenly than the current system. Every tiles the center of the laser passes over, however briefly, is exposed to the laser at some point in the The angling is just an ascetic thing.

**Comment by KazeEspada** [comment]
It should lock the shot location to the center of the clicked tile at least

**Comment by KorPhaeron** [comment]
Jesus christ I can't hit anything I'm having second thoughts about this

**Comment by PKPenguin321** [comment]
I warned you bro
I told you about those janky hitboxes bro

File: code/game/objects/items/stacks/sheets/sheet_types.dm (hunk 1 above)
Comment (optimumtact) [inline code/game/objects/items/stacks/sheets/sheet_types.dm]: Commenting it out is bad, please fix by removing it.

**Comment by optimumtact** [comment]
This would only work if the projectile collisions were pixel as well

**Comment by KorPhaeron** [comment]
It's not the hitboxes, it's having to aim dead center of the tile you want, or else it veers wildly off to the side

**Comment by Fox-McCloud** [comment]
>Jesus christ I can't hit anything I'm having second thoughts about this

Hitboxes are more or less the same as they are before, but it does definitely impact how it's perceived you have to aim. It deffffinitely makes aiming and hitting someone more difficult, but after a week, you get onto it pretty easily.

**Comment by Delimusca** [comment]
looks like the angular rounding errors are still a problem, that's what causes the janky hitboxes on vectors between all octant lines. from what i know about trig, it's some problem with arctan being applied to a grid that needs to be corrected for.

i'd look at how goon did it since theirs seems to work perfectly.

File: code/modules/projectiles/firing.dm [hunk 2]
```diff
@@ -65,4 +57,41 @@
...
+			var/list/screen_loc_X = splittext(screen_loc_params[1],":")
+
+			//Split Y+Pixel_Y up into list(Y, Pixel_Y)
+			var/list/screen_loc_Y = splittext(screen_loc_params[2],":")
+			// world << "X: [screen_loc_X[1]] PixelX: [screen_loc_X[2]] / Y: [screen_loc_Y[1]] PixelY: [screen_loc_Y[2]]"
+			var/x = text2num(screen_loc_X[1]) * 32 + text2num(screen_loc_X[2]) - 32
+			var/y = text2num(screen_loc_Y[1]) * 32 + text2num(screen_loc_Y[2]) - 32
+
+			//Calculate the "resolution" of screen based on client's view and world's icon size. This will work if the user can view more tiles than average.
+			var/screenview = (user.client.view * 2 + 1) * world.icon_size //Refer to http://www.byond.com/docs/ref/info.html#/client/var/view for mad maths
```
Comment (MrPerson) [inline code/modules/projectiles/firing.dm]: Seems an awful lot like you could just use client.bounds to figure this out

File: code/modules/projectiles/projectile.dm [hunk 3]
```diff
@@ -119,23 +133,83 @@
...
-				if(loc == get_turf(original))
-					if(!(original in permutated))
-						Bump(original, 1)
-			Range()
-		sleep(1)
+/obj/item/projectile/proc/fire(var/setAngle)
+	if(setAngle)
+		Angle = setAngle
+	if(!legacy) //new projectiles
+		set waitfor = 0
```
Comment (MrPerson) [inline code/modules/projectiles/projectile.dm]: nope

File: code/modules/projectiles/projectile.dm [hunk 4]
```diff
@@ -119,23 +133,83 @@
...
+				if(spread)
+					Angle += (rand() - 0.5) * spread
+				var/matrix/M = new
+				M.Turn(Angle)
+				transform = M
+
+				var/Pixel_x=round(sin(Angle)+16*sin(Angle)*2)
+				var/Pixel_y=round(cos(Angle)+16*cos(Angle)*2)
+				var/pixel_x_offset = pixel_x + Pixel_x
+				var/pixel_y_offset = pixel_y + Pixel_y
```
Comment (MrPerson) [inline code/modules/projectiles/projectile.dm]: You're going to name the var Pixel_x? Come on now. I shouldn't have to explain the perils of having an object var named pixel_x and a local var named Pixel_x.

**Comment by coiax** [comment]
1. So does this actually change the path of the projectile or just makes it LOOK like it's changed?
2. Will this actually make the firing testing range a legitimate and useful way of testing out the fire patterns and spread of guns and aiming and shit.

**Comment by KazeEspada** [comment]
If you notice in the pic on the left, the circled area is where they clicked. Well, most of them were clicked in a tile center, except the bottom one which wasn't. The bullet chooses to go off to the side because it is following the path to that pixel point rather than locking to the center of the tile clicked on which would make more sense in a tile based enviroment.

**Comment by ChangelingRain** [comment]
Also frankly the projectiles look really jerky and slow instead of smooth.

**Comment by Zilenan91** [comment]
It feels janky at first, going on pixel projectiles, believe me. But after playing on other servers that have them it's honestly so much better once you get used to it.

**Comment by Incoming5643** [comment]
"other servers do this" is not a good arguement

**Comment by KorPhaeron** [comment]
>GOON IS SO GOOD PORT ALL OF GOON
>gloom is empty and dead
>2013 CODE IS SO GOOD REVERT EVERYTHING
>2013 server is empty and dead

Incoming is right

**Comment by Incoming5643** [comment]
to be fair I am also empty and dead

**Comment by Zilenan91** [comment]
Gloom was just something shadowlight threw together because he wanted to after we ran a few rounds of Goon after the leak. 2013 code is/was basically the same thing, it was just done on kind of a whim and isn't necessarily indicative of what people actually want since there was little desire for these servers in the first place.

**Comment by Jalleo** [comment]
What Zilenan91 says is true they want stuff but they dont want to lose the new content or they want it as a special event for just a few rounds then its back to standard code not a full out new server.

**Comment by KorPhaeron** [comment]
Paprika quit ss13 forever again so someone should either merge or deny this since it wont survive conflicts

**Comment by Zilenan91** [comment]
Again huh?

**Comment by lzimann** [comment]
I think we are beter without it tbh

**Comment by Bawhoppen** [comment]
its so much better though

**Comment by Zilenan91** [comment]
I say we merge it and finally get rid of shitty projectile paths

**Comment by Core0verload** [comment]
No merge for this, as it wrecks shooting with shitty hitboxes.

**Comment by Bawhoppen** [comment]
@Core0verload and you are in charge of what?

**Comment by Core0verload** [comment]
No. I just did a test run. It looks cool, but sucks.

**Comment by Bawhoppen** [comment]
@Core0verload I redirect you to my previous comment

File: code/modules/projectiles/guns/energy/stun.dm [hunk 5]
```diff
@@ -50,3 +50,73 @@
 /obj/item/weapon/gun/energy/disabler/cyborg/newshot()
 	..()
 	robocharge()
+
+//A realistic, cartridge-based taser that must use disposable cartridges.
```
Comment (lzimann) [inline code/modules/projectiles/guns/energy/stun.dm]: nice "stealth" feature addition

File: code/game/objects/items/stacks/sheets/sheet_types.dm [hunk 6]
```diff
@@ -34,7 +34,7 @@ var/global/list/datum/stack_recipe/metal_recipes = list ( \
 	new/datum/stack_recipe("firelock frame", /obj/structure/firelock_frame, 3, time = 50, one_per_turf = 1, on_floor = 1), \
 	new/datum/stack_recipe("turret frame", /obj/machinery/porta_turret_construct, 5, time = 25, one_per_turf = 1, on_floor = 1), \
 	new/datum/stack_recipe("meatspike frame", /obj/structure/kitchenspike_frame, 5, time = 25, one_per_turf = 1, on_floor = 1), \
-	new/datum/stack_recipe("reflector frame", /obj/structure/reflector, 5, time = 25, one_per_turf = 1, on_floor = 1), \
+/*	new/datum/stack_recipe("reflector frame", /obj/structure/reflector, 5, time = 25, one_per_turf = 1, on_floor = 1), \*/
```
Comment (GunHog) [inline code/game/objects/items/stacks/sheets/sheet_types.dm]: This here is the only part that I dislike.

File: code/modules/projectiles/guns/energy/stun.dm (hunk 5 above)
Comment (GunHog) [inline code/modules/projectiles/guns/energy/stun.dm]: This should be in its own PR!

**Comment by phil235** [comment]
PR are supposed to be atomized, you're adding pixel projectile and taser cartridge in the same PR, this is bad.

**Comment by KorPhaeron** [comment]
I'm not actually adding anything, I just made a PR straight from paprika's branch so I don't have any control if/when new commits are added

**Comment by GunHog** [comment]
So...you are helping Paprika ban evade via an alt account? I get that this is Paprika's content and all, but why are you not using your own branch? It does not seem to be a good idea to let someone else have commit control over a PR with your name and position tied to it.

**Comment by PKPenguin321** [comment]
Function > form in this case

:-1:

**Comment by KorPhaeron** [comment]
>So...you are helping Paprika ban evade via an alt account?

I said that in the open yeah

**Comment by KorPhaeron** [comment]
@ChangelingRain @phil235 @Cheridan @RemieRichards @Jordie0608 

Someone close or merge this please. The taser thing was apparently an accident, but it's adminspawn only in this PR so it shouldn't be awful, right?

**Comment by lzimann** [comment]
this is going to be awful if it gets merged tbh

**Comment by kilkun** [comment]
The thing about the corners is, it actually means you can fucking take cover. Anything that makes combat more tactical is A++.

**Comment by lzimann** [comment]
just get good and dodge :^)

**Comment by Delimusca** [comment]
is this even maintainable, it's already broken and nobody seems to know why.

**Comment by kilkun** [comment]
sounds like ss13 worthy to me!

**Comment by Bawhoppen** [comment]
just speed up projectiles so they're as fast as old ones

**Comment by KorPhaeron** [comment]
>merge conflicts

I guess it's over

**Comment by lzimann** [comment]
good

**Comment by Bawhoppen** [comment]
don't let another dream die

**Comment by KorPhaeron** [comment]
@ChangelingRain @phil235 @Cheridan @RemieRichards @Jordie0608 

Should I fix this or no?

**Comment by phil235** [comment]
@KorPhaeron I suggest you make your own branch so you can change the code if needed.

**Comment by KorPhaeron** [comment]
I will if I know this is gonna get merged, I don't want to deal with it sitting in limbo though

**Comment by KorPhaeron** [comment]
It's up to bawhoppen now
